### PR TITLE
Add cram tests for diffing changes to types

### LINKED ISSUE #8: Add test cases for diffing changes to types
Having a few simple test cases for different API changes would make for a great way to monitor progress on `api-diff` initial implementation.

Add tests that will check the behavior of `api-diff` when

- [ ] Adding a type
- [ ] Removing a type
- [ ] Modifying an existing type

Those tests should be added to `tests/api-diff/run.t`. We can reuse the base `ref.cmi` as the reference interface. For each test case we should generate a new `.mli` file with a change from the list above, compile it to a `.cmi` with `ocamlc` and then run `api-diff ref.cmi <test-file.cmi>`.

You can read about the testing methodology used for the tests at https://github.com/NathanReb/ocaml-api-watch/blob/main/CONTRIBUTING.md#writing-tests

### PR DESCRIPTION
**Pull request by MungaSoftwiz**
This draft pull request is for the issue 8. If merged it Closes #8 .
I am requesting a review on this draft PR because I encounter some errors listed below: @NathanReb @shonfeder @dinakajoy 

```
File "tests/api-diff/run.t", line 1, characters 0-0:
diff --git a/_build/.sandbox/721d8d27ea32e21877549a55015d08cf/default/tests/api-diff/run.t b/_build/.sandbox/721d8d27ea32e21877549a55015d08cf/default/tests/api-diff/run.t.corrected
index c2ff27f..a4a1199 100644
--- a/_build/.sandbox/721d8d27ea32e21877549a55015d08cf/default/tests/api-diff/run.t
+++ b/_build/.sandbox/721d8d27ea32e21877549a55015d08cf/default/tests/api-diff/run.t.corrected
@@ -44,11 +44,17 @@ Here we generate a basic `.mli` files with two types and a function:
 We generate the .cmi file

   $ ocamlc add_type.mli
+  File "add_type.mli", line 5, characters 0-0:
+  Error: Syntax error
+  [2]

Now we run api-watcher on the two cmi files, there should be a difference:

   $ api-diff ref.cmi add_type.cmi
-  API changed!
+  api-watcher: CMI argument: no 'add_type.cmi' file or directory
+  Usage: api-watcher [OPTION]… CMI CMI
+  Try 'api-watcher --help' for more information.
+  [124]
```

```
### A file with a removed type:

@@ -78,4 +84,7 @@ We generate a .cmi file
 Now we run api-watcher on the two cmi files, there should be a difference:

   $ api-diff ref.cmi modify_type.cmi
-  API changed!
+  api-watcher: internal error, uncaught exception:
+               Includemod.Error(_)
+
+  [125]
```
Also the file with a removed type doesn't display anything.

### CODE REVIEW

**Comment by harshey1103** [comment]
Hi @MungaSoftwiz, I have encountered a similar error. You can refer to https://github.com/NathanReb/ocaml-api-watch/pull/14#issuecomment-1981674445 this thread.

File: tests/api-diff/run.t
Comment (harshey1103) [inline tests/api-diff/run.t:42]: Also, please don't leave any whitespaces. Here, whitespace after EOF is contributing to syntax error which hinders the compilation. Hence you are receiving this error:
```
+  File "add_type.mli", line 5, characters 0-0:
+  Error: Syntax error
+  [2]

Now we run api-watcher on the two cmi files, there should be a difference:

   $ api-diff ref.cmi add_type.cmi
-  API changed!
+  api-watcher: CMI argument: no 'add_type.cmi' file or directory
+  Usage: api-watcher [OPTION]… CMI CMI
+  Try 'api-watcher --help' for more information.
+  [124]
```
These cram tests are very specific about the syntax :p

**Comment by MungaSoftwiz** [comment]
> Hi @MungaSoftwiz, I have encountered a similar error. You can refer to [#14 (comment)](https://github.com/NathanReb/ocaml-api-watch/pull/14#issuecomment-1981674445) this thread.

Thank you @harshey1103. I saw the thread but the issue is I have more errors in addition to the one that you encountered. That's why I am still not sure how to go about it. For the one about `Includemod.Error(_)` I see it's an error in the code.

**Comment by NathanReb** [comment]
The first error you have indicates that the compiler failed to produce the `.cmi` file because your `.mli` file as a syntax error.

As @harshey1103 pointed out, the trailing whitespace after `EOF` may mess up the heredoc.

For debugging you can add an extra `$ cat add_type.mli` before running `ocamlc` to see what actually makes it into the file. Fixing this should fix your syntax error and therefore the following `no add_type.cmi` one.

**Comment by MungaSoftwiz** [comment]
> The first error you have indicates that the compiler failed to produce the `.cmi` file because your `.mli` file as a syntax error.
> 
> As @harshey1103 pointed out, the trailing whitespace after `EOF` may mess up the heredoc.
> 
> For debugging you can add an extra `$ cat add_type.mli` before running `ocamlc` to see what actually makes it into the file. Fixing this should fix your syntax error and therefore the following `no add_type.cmi` one.

The tests work now. I will pushing the commit.

File: tests/api-diff/run.t
Comment (NathanReb) [inline tests/api-diff/run.t:43]: Can you remove this trailing whitespace please?

You probably want to configure your editor to warn you about trailing whitespaces.

File: tests/api-diff/run.t
Comment (NathanReb) [inline tests/api-diff/run.t:58]: Here you are removing everything, not just a type! Can you make it so the only diff between this and the reference is a removed type?

**Review by NathanReb — DISMISSED** [review]

File: tests/api-diff/run.t
Comment (MungaSoftwiz) [reply tests/api-diff/run.t:58]: Okay thank you for the review. Working on that asap!

File: tests/api-diff/run.t
Comment (MungaSoftwiz) [reply tests/api-diff/run.t:43]: Thank you for the suggestion. For anyone who might experience the trailing whitespaces issue too, here is  a script you can use in your editor to highlight, match and remove white spaces before saving to buffer.

```
highlight ExtraWhitespace ctermbg=red guibg=red
match ExtraWhitespace /\s\+$/
autocmd BufWritePre * %s/\s\+$//e
```

**Comment by MungaSoftwiz** [comment]
I updated the reference module interface to add an alias string type so that if it's removed the `Includemod.Error(_)` would detect this. The current Includemod.Error does not detect the removal of the unused type as a change to the API. I'm waiting on the resolve of `Includemod.Error(_)` to be able to proceed with these changes and test. Can you please review @NathanReb @shonfeder.

Here is the exception:
```
   $ api-diff ref.cmi remove_type.cmi
-  api-watcher: internal error, uncaught exception:
-               Includemod.Error(_)
-
-  [125]
+  API unchanged!
```

**Comment by NathanReb** [comment]
Yes I think this is somewhat expected unfortunately and show we can't use `Includemod` for what we're trying to achieve here.

It was used as a place holder until an actual implementation. I suggest you promote this result but mentions in the text above that the tool currently fails to detect this change and that it should be fixed! That way it's clear looking at the test that this is not the expected behaviour of the tool!

Indeed, it's best if you wait until #19 is merged and then rebase your work. I'll ping you here once you can resume work on this!

File: tests/api-diff/run.t
Comment (shonfeder) [inline tests/api-diff/run.t:67]: Let's add a `TODO` prefix here and a link to the new issue I've created to track the problem. This will help us remember to fix it and make it easier to locate:

```suggestion
TODO: The api-diff tool fails to detect this particular change and should be fixed!
See https://github.com/NathanReb/ocaml-api-watch/issues/23
```

**Review by shonfeder — CHANGES_REQUESTED** [review]
Looks great. Awesome work here. I just have two more small suggestions, and we should be good to merge!

File: tests/api-diff/run.t
Comment (MungaSoftwiz) [reply tests/api-diff/run.t:34]: I tried using this method to sort of decouple it from the line 14 comment. You see for line 14, it checks for two identical types. I did the markdown syntax on the comments to be able to differentiate type tests of the same file and type tests of two different files. I think if mixed together, the test file could be less readable. So it was a step to enhance readability. @shonfeder

Comment (shonfeder) [reply tests/api-diff/run.t:34]: Sorry I wasn't clear: my question was just about the line 

```
Generate a basic `.mli` files with two types and a function
```

This is not associated with anything in particular as far as can tell. The markdown-like formatting is fine IMO, and there is precedent for that in this file, but this line seems out of place.

Comment (MungaSoftwiz) [reply tests/api-diff/run.t:34]: This line came with the original `ref.mli` as an explanation of creating a base file that has a two types and a function. Then from the base file we can describe changes to files. For instance:
```
Generate a file with an added type
```
I also see it says `files` instead of `file`. Please let me know if my understanding is correct, I can remove it if needed.

Comment (shonfeder) [reply tests/api-diff/run.t:34]: That is what I thought. This comment is not relevant in this context tho, as we are not creating the original `ref.mil` here. I suggest just removing this line

```suggestion
```

**Comment by shonfeder** [comment]
One more change will be needed here now that #24 is in, then we should be good to go!

If you merge in the main branch (or rebase on to it) then the test for removing a type should now be fixed. You can update that test and remove the TODO!

File: tests/api-diff/run.t
Comment (MungaSoftwiz) [reply tests/api-diff/run.t:32]: Thanks for the suggestion. I have updated the file, it's ready for review @shonfeder

**Comment by MungaSoftwiz** [comment]
> One more change will be needed here now that #24 is in, then we should be good to go!
> 
> If you merge in the main branch (or rebase on to it) then the test for removing a type should now be fixed. You can update that test and remove the TODO!

Great! Working on it asap!

**Comment by shonfeder** [comment]
I think you can just promote the new results now and then you should be good to go!

**Comment by shonfeder** [comment]
Looks great! I just made one small fix, for a mistake that I think was do to miscommunication on my part.

Thank you!

**Review by shonfeder — APPROVED** [review]

**Comment by MungaSoftwiz** [comment]
> Looks great! I just made one small fix, for a mistake that I think was do to miscommunication on my part.
> 
> Thank you!
Thanks for the review. Looking forward to new issues.

**Pull request merged**
